### PR TITLE
ng_ipv6: add 6in6 encapsulation support

### PR DESCRIPTION
**Pull request by authmillenon**
For sending the "user" is expected just to provide the preinitialized headers according to [RFC2473](http://tools.ietf.org/html/rfc2473).

### CODE REVIEW

**Comment by OlegHahm** [comment]
Is this really that easy? Can you post a pointer to the spec?

**Comment by authmillenon** [comment]
Don't see why it should be more difficult: http://tools.ietf.org/html/rfc2473#section-3.3

**Comment by OlegHahm** [comment]
I see, it seems really easy. Cool. But looking at this PR again, I don't understand what it does. Decapsulating an encapsulated IPv6 packet before sending? What's the purpose?

**Comment by authmillenon** [comment]
Decapsulating. Encapsulating can be done by just putting an IPv6 header in front of a full IPv6 packet, so I did not see it to be necessary to implement this here:

```C
tunnel = ng_ipv6_hdr_build(orig, NULL, 0, &dest, sizeof(dest));
```

**Comment by OlegHahm** [comment]
Ah, sorry, the diff presentation confused me. I thought this was sending code.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 1]
```diff
@@ -53,6 +53,9 @@ static void _send(ng_pktsnip_t *pkt, bool prep_hdr);
 /* Main event loop for IPv6 */
 static void *_event_loop(void *args);
 
+/* Handles encapsulated IPv6 packets */
+static void _handle_ipv6_encapsulation(ng_pktsnip_t *pkt);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: I would prefer a name like `_decapsulation` or `_encapsulated`.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 2]
```diff
@@ -594,5 +600,16 @@ static void _receive(ng_pktsnip_t *pkt)
     ng_ipv6_demux(iface, pkt, hdr->nh);
 }
 
+static void _handle_ipv6_encapsulation(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *ptr = pkt;
+
+    /* Remove encapsulating IPv6 header */
+    while ((ptr->next != NULL) && (ptr->next->type == NG_NETTYPE_IPV6)) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Is it allowed to encapsulate more than one IPv6 packet? If yes, are we sure, we want only the innermost packet here?

Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: That is not what is happening here: its supposed to remove the outer-most IPv6 header + extension headers. But one bug is in there (`pkt->type` was set to `NG_NETTYPE_IPV6` in l73, so it (the encapsulated) is removed here to). Will fix.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 1 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: will do.

**Comment by authmillenon** [comment]
Addressed commends

**Comment by OlegHahm** [comment]
ACK

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 3]
```diff
@@ -594,5 +600,21 @@ static void _receive(ng_pktsnip_t *pkt)
...
 
+static void _decapsulate(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *ptr = pkt;
+
+    pkt->type = NG_NETTYPE_UNDEF;   /* prevent payload (the encapsulated packet)
+                                     * from being removed */
+
+    /* Remove encapsulating IPv6 header */
+    while ((ptr->next != NULL) && (ptr->next->type == NG_NETTYPE_IPV6)) {
```
Comment (cgundogan) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: does this also remove the next headers, which are part of the encapsulating ipv6 header?

Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Yes.

**Comment by cgundogan** [comment]
I do not know how to test this, but the code is reasonable. ACK. Seems like @OlegHahm also has no further objection. Go when travis shows green.

**Comment by cgundogan** [comment]
Ah, squash please (:

**Comment by authmillenon** [comment]
Don't understand why github only notified me about your comments today, but did :-)

**Pull request merged**